platforms/core-execution/execution/src/main/java/org/gradle/internal/execution/steps/AssignImmutableWorkspaceStep.java

```diff
@@ -154,56 +151,49 @@ public WorkspaceResult execute(UnitOfWork work, C context) {
+            AtomicMoveImmutableWorkspace workspace = workspaceProvider.getAtomicMoveWorkspace(uniqueId);
             return loadImmutableWorkspaceIfExists(work, workspace)
                 .orElseGet(() -> executeInTemporaryWorkspace(work, context, workspace));
         }
     }
 
     private void deleteStaleFiles(File workspace) {
-        Path lockFile = DefaultFileLockManager.determineLockTargetFile(workspace).toPath();
-        try (DirectoryStream<Path> paths = Files.newDirectoryStream(workspace.toPath(), path -> !path.equals(lockFile))) {
-            for (Path path : paths) {
-                deleter.deleteRecursively(path.toFile());
-            }
+        try  {
+            deleter.deleteRecursively(workspace);
         } catch (IOException e) {
             throw new UncheckedIOException(e);
         }
     }
 
     private Optional<WorkspaceResult> loadImmutableWorkspaceIfExists(UnitOfWork work, ImmutableWorkspace workspace) {
         File immutableLocation = workspace.getImmutableLocation();
-        // We only want to check if the immutable workspace exists, so filter the content of the directory.
-        // That way we also avoid lock file snapshotting, that can be deleted without notifying fileSystemAccess
-        Optional<FileSystemLocationSnapshot> workspaceSnapshot = fileSystemAccess.read(immutableLocation.getAbsolutePath(), new SkipDirectoryContentFilter());
-        return workspaceSnapshot.flatMap(snapshot -> {
-            switch (snapshot.getType()) {
-                case Directory:
-                    return loadImmutableWorkspaceIfConsistent(work, workspace);
-                case RegularFile:
-                    throw new IllegalStateException(
-                        "Immutable workspace is occupied by a file: " + immutableLocation.getAbsolutePath() + ". " +
-                            "Deleting the file in question can allow the content to be recreated.");
-                case Missing:
-                    return Optional.empty();
-                default:
-                    throw new AssertionError();
-            }
-        });
+        FileSystemLocationSnapshot snapshot = fileSystemAccess.read(immutableLocation.getAbsolutePath());
+        switch (snapshot.getType()) {
+            case Directory:
+                return loadImmutableWorkspaceIfConsistent(work, workspace);
+            case RegularFile:
+                throw new IllegalStateException(
+                    "Immutable workspace is occupied by a file: " + immutableLocation.getAbsolutePath() + ". " +
+                        "Deleting the file in question can allow the content to be recreated.");
+            case Missing:
+                return Optional.empty();
+            default:
+                throw new AssertionError();
+        }
     }
 
     private Optional<WorkspaceResult> loadImmutableWorkspaceIfConsistent(UnitOfWork work, ImmutableWorkspace workspace) {
-        File immutableLocation = workspace.getImmutableLocation();
```

Comment:
🤔 Don't see a reason to inline it. I'd keep it as before

Reply:
Fixed